### DISCUSSION TITLE
Specify working directory XUnit AssemblyRunner.

### DISCUSSION THREAD

**Discussion by TimonPost** [xunit/xunit#2252] [Question] [answered]
I am creating an instance of [AssemblyRunner](https://github.com/xunit/samples.xunit/blob/main/TestRunner/Program.cs) and execute the tests from a `ConsoleApplication`:

_Project Stucture:_
- Console Application
- XUnitTestproject

```csharp
var testPath = Path.Combine(BASE_TEST_DIR, "XUnitTestProject.dll");

using (var runner = AssemblyRunner.WithoutAppDomain(testPath))
{
    runner.OnDiscoveryComplete = OnDiscoveryComplete;
    ...    
    runner.Start();
    ...
}
```

When I run the code, it says: `Could not find/load any of the following assemblies: xunit.execution.dotnet.dll`. This assembly is located in `XUnitTestproject` but not in `Console Application`. If I add the `XUnitTestproject` as a dependency to `Console Application` it does work perfectly fine. 

I know this can be resolved by installing XUnit on the `Console Application`, but in my case, this is not desired. Next to that, I do not want to have a reference to `XUnitTestproject`. To fix this, `dotnet test`, `nunit`, allows me to set a working directory/base directory such that when I set that to the `XUnitTestproject` it will load the required assemblies from there. Is this possible?

**Comment by bradwilson** [comment]
The base folder for dependency resolution should be the one that has `XUnitTestProject.dll` in it by default. Can you provide a repro project? Maybe I can track down the issue.

**Reply by TimonPost** [reply]
There you go: [XUnitIssue.zip](https://github.com/xunit/xunit/files/5976483/XUnitIssue.zip)

**Comment by TimonPost** [comment, marked as answer]
I resolved the issue by handing `AppDomain.CurrentDomain.AssemblyResolve` event and serve it with the correct directory. Apparently, the default location assemblies will be resolved from is the executing application. See this [article ](https://weblog.west-wind.com/posts/2016/dec/12/loading-net-assemblies-out-of-seperate-folders#Checking-already-loaded-Assemblies?) for more info.

```csharp
private static Assembly CurrentDomain_AssemblyResolve(object sender, ResolveEventArgs args)
{
    // Ignore missing resources
    if (args.Name.Contains(".resources"))
        return null;

    // check for assemblies already loaded
    Assembly assembly = AppDomain.CurrentDomain.GetAssemblies().FirstOrDefault(a => a.FullName == args.Name);
    if (assembly != null)
        return assembly;

    // Try to load by filename - split out the filename of the full assembly name
    // and append the base path of the original assembly (ie. look in the same dir)
    string filename = args.Name.Split(',')[0] + ".dll".ToLower();

    string asmFile = Path.Combine(projectDir.FullName, "XUnitTestProject", "bin", "Debug", "netcoreapp3.1", "copy1", filename);

    try
    {
        return System.Reflection.Assembly.LoadFrom(asmFile);
    }
    catch (Exception ex)
    {
        return null;
    }
}
```

And make sure to register the event handler

```charp
AppDomain.CurrentDomain.AssemblyResolve += CurrentDomain_AssemblyResolve;
```

Now, every time an assembly fails to load, because it does not exist in the application directory, you can serve the assembly from the path you know it exists.
